### PR TITLE
[GSoC] Added aseries function

### PR DESCRIPTION
**Pull request by arighnac**
#### Brief description of what is fixed or changed

Added asymptotic expansion of series. This is equivalent to `self.series(x, oo, n)`.
Tests and docs have been added.

**BEFORE**

```
>>> from sympy import *
>>> x=symbols('x')
>>> e = sin(1/x + exp(-x)) - sin(1/x)
>>> e.series(x,oo)
O(x**(-6), (x, oo))
```

**NOW**

```
>>> from sympy import *
>>> x=symbols('x')
>>> e = sin(1/x + exp(-x)) - sin(1/x)
>>> e.aseries(x)
(1/(24*x**4) - 1/(2*x**2) + 1 + O(x**(-6), (x, oo)))*exp(-x)
```

#### Other comments

**NOTE :** - This is a work in progress.

#### Release Notes

* series
   * Implemented `aseries` function in `sympy.core.expr` and `sympy.series.aseries`.

### CODE REVIEW

**Comment by ArighnaIITG** [comment]
This PR is almost ready @leosartaj . Open for reviewing.

**Comment by leosartaj** [comment]
> This PR is almost ready @leosartaj . Open for reviewing.

Can you elaborate what do you mean by almost?

**Comment by ArighnaIITG** [comment]
Two symbolic tests are not passing, i.e., tests in which there are some symbolic terms. The logic is ready, and all the other tests are passing.

File: sympy/core/expr.py [hunk 1]
```diff
@@ -2958,6 +2957,149 @@ def yield_lseries(s):
...
+                    break
+                elif snew.getO():
+                    has_ord = True
+                s += (snew * k**expo)
+            else:
+                s += t
+
+        if not o or has_ord:
+            return s.subs(k, exp(logw))
+        return (s + o).subs(k, exp(logw))
```
Comment (leosartaj) [inline sympy/core/expr.py]: Add a few comments for key steps.

**Comment by ArighnaIITG** [comment]
@leosartaj how do I switch the implementation here? How do I call the `aseries` from `core/expr.py`, keeping the logic of the code in `series/aseries.py`?

**Comment by ArighnaIITG** [comment]
@leosartaj . Comments added. Also, I felt that keeping the same code structure would be better, because if we shift the logic to the `aseries.py` file, that becomes a separate function now. There are a lot of recursive calls which then would have to be changed. It would only create a lot more errors. Also, since `series` and other sub-methods of `series` are maintained in the same exact way, I feel that consistency will also be maintained.

File: sympy/core/expr.py [hunk 2]
```diff
@@ -2849,12 +2848,10 @@ def series(self, x=None, x0=0, n=6, dir="+", logx=None):
...
-            sgn = 1 if x0 is S.Infinity else -1
-            s = self.subs(x, sgn/x).series(x, n=n, dir='+')
-            if n is None:
-                return (si.subs(x, sgn/x) for si in s)
-            return s.subs(x, sgn/x)
-
+            s = self.aseries(x, n)
+            if x0 is S.NegativeInfinity:
+                return s.subs(x, -x)
+            return s
```
Comment (leosartaj) [inline sympy/core/expr.py]: I am not sure about this. Let's not change the default for now. We can integrate this later. We need to be sure ``aseries`` will cover all the cases which ``series`` covers also maintaining correctness.

File: sympy/core/expr.py [hunk 3]
```diff
@@ -2958,6 +2955,154 @@ def yield_lseries(s):
...
+            xpos = Dummy('x', positive=True)
+            return self.subs(x, xpos).aseries(xpos, n, bound, hir).subs(xpos, x)
+
+        om, exps = mrv(self, x)
+
+        """
+        We move one level up by replacing `x` by `exp(x)`, and then
+        computing the asymptotic series for f(exp(x)). Then asymptotic series
+        can be obtained by moving one-step back, by replacing x by ln(x).
+        """
```
Comment (leosartaj) [inline sympy/core/expr.py]: I can we make this multi-line comment (using #) rather than docstrings.

File: sympy/core/expr.py [hunk 4]
```diff
@@ -2958,6 +2955,154 @@ def yield_lseries(s):
...
+        if hir:
+            return s.subs(k, exp(logw))
+
+        o = s.getO()
+        terms = sorted(Add.make_args(s.removeO()), key=lambda i: int(i.as_coeff_exponent(k)[1]))
+        s = S.Zero
+        has_ord = False
+
+        """Then we recursively expand these coefficients one by one into
+        their asymptotic series in terms of their most rapidly varying subexpressions."""
```
Comment (leosartaj) [inline sympy/core/expr.py]: Use # for comments.

File: sympy/core/expr.py [hunk 5]
```diff
@@ -2958,6 +2955,154 @@ def yield_lseries(s):
...
+        for t in terms:
+            coeff, expo = t.as_coeff_exponent(k)
+            if coeff.has(x):
+                """Recursive step."""
```
Comment (leosartaj) [inline sympy/core/expr.py]: #

File: sympy/series/aseries.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    .. [1] A New Algorithm for Computing Asymptotic Series - Dominik Gruntz
+    .. [2] Gruntz thesis - p90
+    .. [3] http://en.wikipedia.org/wiki/Asymptotic_expansion
+
+    See Also
+    ========
+
+    See the docstring of Expr.aseries() for complete details of this wrapper.
+
+    """
```
Comment (leosartaj) [inline sympy/series/aseries.py]: Let's maintain docstring at one place rather than two copies of it.

**Comment by leosartaj** [comment]
Can you give an example where series fails and this works?

File: sympy/core/expr.py [hunk 7]
```diff
@@ -9,7 +9,6 @@
 from .compatibility import reduce, as_int, default_sort_key, range, Iterable
 from sympy.utilities.misc import func_name
 from mpmath.libmp import mpf_log, prec_to_dps
-
```
Comment (leosartaj) [inline sympy/core/expr.py]: Let's keep this line.

**Review by leosartaj — CHANGES_REQUESTED** [review]

**Comment by ArighnaIITG** [comment]
> Can you give an example where series fails and this works?

The example has been stated in the PR description. All the test cases mentioned in the `tests/test_aseries.py` return only an order term, which is completely wrong. After implementation of this function, they return the asymptotic series.

**Comment by ArighnaIITG** [comment]
All changes have been integrated @leosartaj .

**Review by leosartaj — APPROVED** [review]

**Comment by leosartaj** [comment]
LGTM. Will merge in 24hrs if no objection.

**Pull request merged**